File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.js
function solution(bridge_length, weight, truck_weights) {
    var time = 0;
    const passing_q = [];      // 다리 위에 있는 트럭 q. [weight, remain_weight]
    let remain_w = weight;
    
    // 지나가고 있는 트럭 q가 빌 때 까지.
    while(passing_q.length || truck_weights.length){
        // console.log("time : ", time, "passing_q : ", passing_q)
        
        if(truck_weights[0] <= remain_w){
            remain_w-=truck_weights[0];
            passing_q.push({weight:truck_weights.shift(), pushTime:time})
        }
        ++time;
        // 경과 시간 - 다리에 올라간 시간이 bridge_length보다 크면 빠져나온다.
        if((time - passing_q[0].pushTime) >= bridge_length){
            remain_w += passing_q[0].weight;
            passing_q.shift();
        }
    }
    ++time;
    return time;
}

File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.cs
using System;
using System.Collections.Generic;
public class Solution {
    public int solution(int bridge_length, int weight, int[] truck_weights) {
        int passedTime = 0;
        int remainWeight = weight;
        Queue<int> readyQ = new Queue<int>(truck_weights);
        // key : weight, value : enqueueTime 
        Queue<KeyValuePair<int, int>> passingQ = new Queue<KeyValuePair<int, int>>();
        
        while(readyQ.Count > 0 || passingQ.Count > 0)
        {
            Console.WriteLine($"passedTime : {passedTime} remainWeight : {remainWeight} readyQ.Count : {readyQ.Count} passingQ.Count : {passingQ.Count} remainWeight : {remainWeight}");
            if(readyQ.TryPeek(out var truckWeight) && remainWeight > truckWeight)
            {
                remainWeight -= truckWeight;
                readyQ.Dequeue();
                passingQ.Enqueue(new KeyValuePair<int, int>(truckWeight, passedTime));
            }
            ++passedTime;
            if(passingQ.TryDequeue(out var passingTruck))
            {
                Console.WriteLine($"?? passingTruck.Value {passingTruck.Value} passedTime : {passedTime} bridge_length : {bridge_length}");
                if(bridge_length <= (passedTime - passingTruck.Value))
                {
                    remainWeight += passingTruck.Key;
                    passingQ.Dequeue();    
                }
                
            }
            if(passedTime > 10)
                break;
        }
        ++passedTime;
        return passedTime;
    }
}